File: source/quantification/quantification_T1.m
% set constants
M0a = 1;
M0b = 0.1;
R   = 15; % s-1
T1a = 1.1 ; % s
T1b = 0.2 ; % s

Ma_ini = -M0a;
Mb_ini = 0.9*M0b;

Kab = R*M0b;
Kba = R*M0a;
R1a = 1/T1a ; % s-1
R1b = 1/T1b ; % s-1

% define model

MatrixArelax = [-(R1a + Kab) ,  Kba ; ...
             Kab , - (R1b + Kba) ];

MatrixArelaxNoExch = [-(R1a) ,  0 ; ...
             0 , - (R1b) ];

MatrixB = [R1a*M0a ; R1b*M0b ];

% relaxation and exchange during time_inc
time_inc = 0.005;
MatrixArelax_exp=expm(MatrixArelax*time_inc);
MatrixABrelax_cst= MatrixArelax\(MatrixArelax_exp-eye(2))*MatrixB;

% relaxation only during time_inc
MatrixArelaxNoExch_exp=expm(MatrixArelaxNoExch*time_inc);
MatrixABrelaxNoExch_cst= MatrixArelaxNoExch\(MatrixArelaxNoExch_exp-eye(2))*MatrixB;

% initialize
time_max = 5;
time_1D = 0:time_inc:time_max;
Np = length(time_1D);

Mdata = zeros(2,Np);
MdataNoExch = zeros(2,Np);

Mini = [Ma_ini ; Mb_ini];  
Mdata(:,1) = Mini;
MdataNoExch(:,1) = Mini;

% Simulation over time
for ii = 2:Np

    Mdata(:,ii) = MatrixArelax_exp*Mdata(:,ii-1) + MatrixABrelax_cst ;
    MdataNoExch(:,ii) = MatrixArelaxNoExch_exp*MdataNoExch(:,ii-1) + MatrixABrelaxNoExch_cst ;

end

% Plot results
h=figure('OuterPosition',[100 50 1200 600]);

subplot(1,2,1)
plot(time_1D,squeeze(Mdata(1,:)),'b',time_1D,squeeze(Mdata(2,:)),'r',time_1D,squeeze(MdataNoExch(1,:)),'b--',time_1D,squeeze(MdataNoExch(2,:)),'r--','LineWidth',2);
xlabel('time (s)');
ylabel('Magnetization (a.u.)');
legend('Ma','Mb','Ma NoExch','Mb NoExch','Location','best')
grid on
box on

subplot(1,2,2)
plot(time_1D,squeeze(Mdata(1,:))/M0a,'b',time_1D,squeeze(Mdata(2,:))/M0b ,'r','LineWidth',2);
xlabel('time (s)');
ylabel('Normalized Magnetization (a.u.)');
legend('Ma/M0a','Mb/M0b','Location','best')
grid on
box on

title(['Input: T1a = ', num2str(1000/R1a),' ms, T1b = ', num2str(1000/R1b),' ms & R = ',num2str(R),' s-1']); 


%% Make noisy signal

% intravoxel averaging, second pool not detectable - short T2b
Signal = Mdata(1,:)/M0a;

% intravoxel averaging, both pools are detectable, neglect T2w
% Signal = Mdata(1,:) + Mdata(2,:);

% set noise properties
noise_mean = 0;
noise_sd = 0.01;

% calculate signal
Signal_meas = Signal + random('normal',noise_mean,noise_sd,1,Np);

% plot results
figure;
plot(time_1D,squeeze(Signal_meas),'x','LineWidth',2);
xlabel('time (s)');
ylabel('Signal (a.u.)');
grid on
box on


%% Simulate data with known analytical solution

R1_short = (R1a + R1b + Kab + Kba + sqrt((R1a - R1b + Kab - Kba)^2 + 4*Kab*Kba))/2;
R1_long = (R1a + R1b + Kab + Kba - sqrt((R1a - R1b + Kab - Kba)^2 + 4*Kab*Kba))/2;
b_short = ((Ma_ini/M0a - 1)*(R1a - R1_long) + (Ma_ini/M0a - Mb_ini/M0b)*Kab)/(R1_short-R1_long);
b_long = -((Ma_ini/M0a - 1)*(R1a - R1_short) + (Ma_ini/M0a - Mb_ini/M0b)*Kab)/(R1_short-R1_long);


Signal_simulated = 1 + b_short*exp(-R1_short*time_1D) + b_long*exp(-R1_long*time_1D);


disp('______________________________________')
disp('Theoretical parameters : ')
disp('______________________________________')
disp([' b_short_theory = ', num2str(b_short)])
disp([' T1_short_theory = ', num2str(1000/R1_short), ' ms'])
disp([' b_long_theory = ', num2str(b_long)])
disp([' T1_long_theory = ', num2str(1000/R1_long), ' ms'])


Rex_pure = Kab + Kba ;
disp([' T1ex_pure = ', num2str(1000/Rex_pure), ' ms'])
disp('______________________________________')


%%  Fit the data

% define fitting function
F = @(x,xdata)(1+ x(1)*exp(-x(2)*xdata)) + x(3)*exp(-x(4)*xdata);


% set intitial guess
x0 = [b_short R1_short b_long R1_long];

% set lower and upper boundaries 
lb = [-10 0 -10 0]; 
ub = [10 1e6 10 1e6];

% perform fitting
tic
[x,resnorm,residual,exitflag,output,lambda,jacobian] = lsqcurvefit(F,x0,time_1D,Signal_meas,lb,ub);
toc

% calculate standard deviation from the covariance matrix of the last fit
% ref : https://comp.soft-sys.matlab.narkive.com/Jq6b64k7/errors-of-parameters-deduced-using-lsqcurvefit
% https://fr.mathworks.com/matlabcentral/answers/465557-calculating-covariance-matrix-from-jacobian-using-lsqcurvefit
% https://stats.stackexchange.com/questions/231868/relation-between-covariance-matrix-and-jacobian-in-nonlinear-least-squares
% mean square error from resnorm, accout for Ndata and Nparam
mse = resnorm /(size(jacobian,1)-size(jacobian,2)); 
CoVMat = inv(jacobian'*jacobian)*mse;
Sd_fromfit = sqrt(diag(CoVMat));

% % AUTRE SOLUTION avec nlinfit au lieu de lsqcurvefit
% % plus simple pour le calcul de la covariance, mais nlinfit may require statistical toolbox
% [beta,R,J,CovB,MSE,] = nlinfit(time_1D,Signal_meas,F,x0);
% Sd_fromfit2 = sqrt(diag(CovB));

if x(4)<x(2)
    bs_est = x(1);
    T1s_est_ms = 1000/x(2);
    bl_est = x(3);
    T1l_est_ms = 1000/x(4);

    bs_sd = Sd_fromfit(1,1);
    T1s_sd_ms = 1000*Sd_fromfit(2,1)/(x(2).^2);
    bl_sd = Sd_fromfit(3,1);
    T1l_sd_ms = 1000*Sd_fromfit(4,1)/(x(4).^2);

else
    bs_est = x(3);
    T1s_est_ms = 1000/x(4);
    bl_est = x(1);
    T1l_est_ms = 1000/x(2);

    bs_sd = Sd_fromfit(3,1);
    T1s_sd_ms = 1000*Sd_fromfit(4,1)/(x(4).^2);
    bl_sd = Sd_fromfit(1,1);
    T1l_sd_ms = 1000*Sd_fromfit(2,1)/(x(2).^2);
end





% plot results
hold on
plot(time_1D,F(x,time_1D),'r','LineWidth',2)
plot(time_1D,squeeze(Signal_simulated),'k--','LineWidth',2);
hold off
title(['Estimated T1S = ', num2str(T1s_est_ms),' ms & T1L = ', num2str(T1l_est_ms),' ms']); 
legend('Meas', 'Fit', 'Theory','Location','best')


disp('______________________________________')
disp('Fitting Results : ')
disp('______________________________________')
disp(['b_short_fit = ', num2str(bs_est), ' +/- ',num2str(bs_sd), ' a.u.'])
disp(['T1_short_fit = ', num2str(T1s_est_ms), ' +/- ',num2str(T1s_sd_ms), ' ms'])
disp(['b_long_fit = ', num2str(bl_est), ' +/- ',num2str(bl_sd), ' a.u.'])
disp(['T1_long_fit = ', num2str(T1l_est_ms), ' +/- ',num2str(T1l_sd_ms),' ms'])
disp('______________________________________')


%% Monte Carlo

% set number of trials
Nbet = 300;

% initilize variable storage
T1s_1D = zeros(1,Nbet);
Bs_1D = zeros(1,Nbet);
T1l_1D = zeros(1,Nbet);
Bl_1D = zeros(1,Nbet);

% define fitting function
F = @(x,xdata)(1+ x(1)*exp(-x(2)*xdata)) + x(3)*exp(-x(4)*xdata);

% set intitial guess
x0 = [b_short R1_short b_long R1_long];

% set lower and upper boundaries 
lb = [-10 0 -10 0]; 
ub = [10 1e6 10 1e6];

h = waitbar(0,' Fitting is still running  : Please wait...'); %,'Position', [ 10 300 250 50]);

tic

for ii = 1 : Nbet
   
    waitbar(ii/Nbet,h)

    % calculate signal
    Signal_meas = Signal + random('normal',noise_mean,noise_sd,1,Np);
    
    % perform fitting
    [x,resnorm,residual,exitflag,output,lambda,jacobian] = lsqcurvefit(F,x0,time_1D,Signal_meas,lb,ub);
    
    % store results
    Bs_1D(ii) = x(1);
    T1s_1D(ii) = 1/x(2);
    Bl_1D(ii) = x(3);
    T1l_1D(ii) = 1/x(4);
        
end

toc

close (h)

% compute statistics over the Nbet trials

bs_mean = mean(Bs_1D);
bs_std = std(Bs_1D);

T1s_mean_ms = 1000*mean(T1s_1D);
T1s_std_ms = 1000*std(T1s_1D);

bl_mean = mean(Bl_1D);
bl_std = std(Bl_1D);

T1l_mean_ms = 1000*mean(T1l_1D);
T1l_std_ms = 1000*std(T1l_1D);

disp('______________________________________')
disp('______________________________________')
disp('Monte Carlo Results : ')
disp('______________________________________')
disp(['bs = ', num2str(bs_mean), ' +/- ',num2str(bs_std) ,' a.u.'])
disp(['T1s = ', num2str(T1s_mean_ms), ' +/- ',num2str(T1s_std_ms) ,' ms'])
disp(['bl = ', num2str(bl_mean), ' +/- ',num2str(bl_std) ,' a.u.'])
disp(['T1l = ', num2str(T1l_mean_ms), ' +/- ',num2str(T1l_std_ms) ,' ms'])
disp('______________________________________')

%% compare with estimates and SD from the last fit

% calculate standard deviation from the covariance matrix of the last fit
mse = resnorm /(size(jacobian,1)-size(jacobian,2)); 
CoVMat = inv(jacobian'*jacobian)*mse;
Sd_fromfit = sqrt(diag(CoVMat));

bs_est = x(1);
T1s_est_ms = 1000/x(2);
bl_est = x(3);
T1l_est_ms = 1000/x(4);

bs_sd = Sd_fromfit(1,1);
T1s_sd_ms = 1000*Sd_fromfit(2,1)/(x(2).^2);
bl_sd = Sd_fromfit(3,1);
T1l_sd_ms = 1000*Sd_fromfit(4,1)/(x(4).^2);

disp('______________________________________')
disp('______________________________________')
disp('Sd from last Fit Results : ')
disp('______________________________________')
disp(['b_short_fit = ', num2str(bs_est), ' +/- ',num2str(bs_sd), ' a.u.'])
disp(['T1_short_fit = ', num2str(T1s_est_ms), ' +/- ',num2str(T1s_sd_ms), ' ms'])
disp(['b_long_fit = ', num2str(bl_est), ' +/- ',num2str(bl_sd), ' a.u.'])
disp(['T1_long_fit = ', num2str(T1l_est_ms), ' +/- ',num2str(T1l_sd_ms),' ms'])
disp('______________________________________')


%% display results
figure;

Nbins = 30 ;

subplot(2,2,1)
histogram(T1s_1D,Nbins)
title(['T1s = ', num2str(T1s_mean_ms),' +/- ', num2str(T1s_std_ms),' ms']); 
xlabel('time (s)');
ylabel('N');

subplot(2,2,3)
histogram(Bs_1D,Nbins)
title(['bs = ', num2str(bs_mean), ' +/- ',num2str(bs_std) ,' a.u.']); 
xlabel('time (s)');
ylabel('N');

subplot(2,2,2)
histogram(T1l_1D,Nbins)
title(['T1l = ', num2str(T1l_mean_ms),' +/- ', num2str(T1l_std_ms),' ms']); 
xlabel('time (s)');
ylabel('N');

subplot(2,2,4)
histogram(Bl_1D,Nbins)
title(['bl = ', num2str(bl_mean), ' +/- ',num2str(bl_std) ,' a.u.']); 
xlabel('time (s)');
ylabel('N');

File: quantification/quantification_T1.py
import matplotlib.pyplot
import numpy
import scipy

## Define and visualize bi-exponential model

# Define species
M0a = 1
M0b = 0.1
R = 15 # s-1
T1a = 1.1 # s
T1b = 0.2 # s

Kab = R*M0b
Kba = R*M0a
R1a = 1/T1a # s-1
R1b = 1/T1b # s-1

# Define model
A = numpy.array([
    [-(R1a + Kab),        Kba],
    [        Kab, -(R1b + Kba)]])

ANoExch = numpy.array([
    [-R1a,    0],
    [   0, -R1b]])

B = numpy.array([R1a*M0a, R1b*M0b])

# relaxation and exchange during time step
step = 0.005 # s
A_exp = scipy.linalg.expm(A*step)
AB_cst = scipy.linalg.inv(A) @ (A_exp-numpy.identity(2)) @ B

# relaxation only during time_inc
ANoExch_exp = scipy.linalg.expm(ANoExch*step)
ABNoExch_cst = scipy.linalg.inv(ANoExch) @ (ANoExch_exp-numpy.identity(2)) @ B

# initialize
time_max = 5
time_1D = numpy.arange(0, time_max+step, step)
Np = len(time_1D)

Mdata = numpy.zeros((2, Np))
MdataNoExch = numpy.zeros((2, Np))

Ma_ini = -M0a
Mb_ini = 0.9*M0b

Mini = numpy.array([Ma_ini, Mb_ini])
Mdata[:,0] = Mini
MdataNoExch[:,0] = Mini

# Simulate over time
for i in range(1, Np):
    Mdata[:,i] = A_exp @ Mdata[:, i-1] + AB_cst
    MdataNoExch[:,i] = ANoExch_exp @ MdataNoExch[:, i-1] + ABNoExch_cst

# Plot results
figure, plots = matplotlib.pyplot.subplots(
    1, 2, sharex=True, sharey=True, layout="tight", figsize=(8,4))
plots[0].plot(time_1D, Mdata[0, :], "b", label="$M_a$")
plots[0].plot(time_1D, Mdata[1, :], "r", label="$M_b$")
plots[0].plot(time_1D, MdataNoExch[0, :], "b--", label="$M_a$ (no exch.)")
plots[0].plot(time_1D, MdataNoExch[1, :], "r--", label="$M_b$ (no exch.)")
plots[0].set(xlabel="Time (s)", ylabel="Magnetization (a.u.)")
plots[0].legend()

plots[1].plot(time_1D, Mdata[0,:]/M0a, "b", label="$M_a/M0_a$")
plots[1].plot(time_1D, Mdata[1, :]/M0b, "r", label="$M_b/M0_b$")
plots[1].set(xlabel="Time (s)")
plots[1].legend()

figure.suptitle(
    f"Input: $T_1^a$ = {1000/R1a:.0f} ms, $T_1^b$ = {1000/R1b:.0f} ms, "
    f"R = {R} s$^{{-1}}$")

figure.savefig("quantification/quantification_T1_BM.png")

## Make noisy signal

# intravoxel averaging, second pool not detectable - short T2b
signal = Mdata[0, :]/M0a

# intravoxel averaging, both pools are detectable, neglect T2w
# signal = Mdata.sum(axis=0)

# set noise properties
noise_mean = 0
noise_sd = 0.01

# calculate signal
signal_meas = signal + numpy.random.normal(noise_mean, noise_sd, signal.shape)

# plot results
figure, plot = matplotlib.pyplot.subplots(layout="tight", figsize=(5, 4))
plot.plot(time_1D, signal_meas, "k")
plot.set(xlabel="Time (s)", ylabel="Signal (a.u.)")

figure.savefig("quantification/quantification_T1_BM_noisy.png")

## Simulate data with known analytical solution

R1_short = (R1a + R1b + Kab + Kba + numpy.sqrt((R1a - R1b + Kab - Kba)**2 + 4*Kab*Kba))/2
R1_long = (R1a + R1b + Kab + Kba - numpy.sqrt((R1a - R1b + Kab - Kba)**2 + 4*Kab*Kba))/2
b_short = ((Ma_ini/M0a - 1)*(R1a - R1_long) + (Ma_ini/M0a - Mb_ini/M0b)*Kab)/(R1_short-R1_long)
b_long = -((Ma_ini/M0a - 1)*(R1a - R1_short) + (Ma_ini/M0a - Mb_ini/M0b)*Kab)/(R1_short-R1_long)

signal_simulated = 1 + b_short*numpy.exp(-R1_short*time_1D) + b_long*numpy.exp(-R1_long*time_1D)

print("Theoretical parameters:")
print(f"  - b_short = {b_short:.3f}")
print(f"  - T1_short = {1000/R1_short:.3f} ms")
print(f"  - b_long = {b_long:.3f}")
print(f"  - T1_long = {1000/R1_long:.3f} ms")

Rex_pure = Kab + Kba
print(f"  - T1ex_pure = {1000/Rex_pure:.3f} ms")

##  Fit the data

# define fitting function
def F(time, b_short, R1_short, b_long, R1_long):
    return (1+ b_short*numpy.exp(-R1_short*time)) + b_long*numpy.exp(-R1_long*time)

# set intitial guess
x0 = [b_short, R1_short, b_long, R1_long]

# set lower and upper boundaries 
lb = [-10, 0, -10, 0] 
ub = [10, 1e6, 10, 1e6]

# perform fitting
x, covariance = scipy.optimize.curve_fit(F, time_1D, signal_meas, x0, bounds=(lb, ub))

# calculate standard deviation from the covariance matrix of the last fit
Sd_fromfit = numpy.sqrt(numpy.diag(covariance))

if x[3] > x[1]:
    # Swap short and long parameters
    x[0], x[2] = x[2], x[0]
    x[1], x[3] = x[3], x[1]

bs_est = x[0]
T1s_est_ms = 1000/x[1]
bl_est = x[2]
T1l_est_ms = 1000/x[3]

# plot results
figure, plot = matplotlib.pyplot.subplots(layout="tight", figsize=(5, 4))
plot.plot(time_1D, F(time_1D, *x), "r", label="Fit")
plot.plot(time_1D, signal_simulated, "k--", label="Simulated")
figure.suptitle(f"Est. $T_1^S$ = {T1s_est_ms:.0f} ms, $T_1^L$ = {T1l_est_ms:.0f} ms")
plot.legend()

figure.savefig("quantification/quantification_T1_fit.png")

print("Fitted parameters:")
print(f"  - b_short = {bs_est:.3f} ± {Sd_fromfit[0]:.3f}")
print(f"  - T1_short = {T1s_est_ms:.3f} ± {1000*Sd_fromfit[1]/(x[1]**2):.3f} ms")
print(f"  - b_long = {bl_est:.3f} ± {Sd_fromfit[2]:.3f}")
print(f"  - T1_long = {T1l_est_ms:.3f} ± {1000*Sd_fromfit[3]/(x[3]**2):.3f} ms")

## Monte Carlo

# set number of trials
Nbet = 300

# initilize variable storage
T1s_1D = numpy.zeros(Nbet)
Bs_1D = numpy.zeros(Nbet)
T1l_1D = numpy.zeros(Nbet)
Bl_1D = numpy.zeros(Nbet)

for i in range(Nbet):
    # calculate signal
    signal_meas = signal + numpy.random.normal(noise_mean, noise_sd, len(signal))
    
    # perform fitting
    x, _ = scipy.optimize.curve_fit(F, time_1D, signal_meas, x0, bounds=(lb, ub))
    
    # store results
    Bs_1D[i] = x[0]
    T1s_1D[i] = 1/x[1]
    Bl_1D[i] = x[2]
    T1l_1D[i] = 1/x[3]

# compute statistics over the Nbet trials

bs_mean = numpy.mean(Bs_1D)
bs_std = numpy.std(Bs_1D)

T1s_mean_ms = 1000*numpy.mean(T1s_1D)
T1s_std_ms = 1000*numpy.std(T1s_1D)

bl_mean = numpy.mean(Bl_1D)
bl_std = numpy.std(Bl_1D)

T1l_mean_ms = 1000*numpy.mean(T1l_1D)
T1l_std_ms = 1000*numpy.std(T1l_1D)

print("Monte Carlo parameters")
print(f"  - b_short = {bs_mean:.3f} ± {bs_std:.3f}")
print(f"  - T1_short = {T1s_mean_ms:.3f} ± {T1s_std_ms:.3f} ms")
print(f"  - b_long = {bl_mean:.3f} ± {bl_std:.3f}")
print(f"  - T1_long = {T1l_mean_ms:.3f} ± {T1l_std_ms:.3f}  ms")

## display results
Nbins = 30
figure, plots = matplotlib.pyplot.subplots(2, 2, layout="tight", figsize=(8,8))

plots[0,0].hist(T1s_1D, Nbins, label=f"$T_1^s$={T1s_mean_ms:.0f} ± {T1s_std_ms:.3f} ms")
plots[0,0].set(xlabel="$T_1^s$ (s)", yticks=[])

plots[1,0].hist(Bs_1D, Nbins, label=f"$b^s$={bs_mean:.3f} ± {bs_std:.3f}")
plots[1,0].set(xlabel="$b^s$ (a.u.)", yticks=[])

plots[0,1].hist(T1l_1D, Nbins, label=f"$T_1^l$={T1l_mean_ms:.0f} ± {T1l_std_ms:.3f} ms")
plots[0,1].set(xlabel="$T_1^l$ (s)", yticks=[])

plots[1,1].hist(Bl_1D, Nbins, label=f"$b^l$={bl_mean:.3f} ± {bl_std:.3f}")
plots[1,1].set(xlabel="$b^l$ (a.u.)", yticks=[])

for plot in plots.ravel():
    plot.legend()
    plot.spines[["left", "top", "right"]].set_visible(False)

figure.savefig("quantification/quantification_T1_monte_carlo.png")
